cpp/src/strings/translate.cu

```diff
@@ -0,0 +1,137 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/strings/string_view.cuh>
+#include <cudf/strings/translate.hpp>
+#include <strings/utilities.hpp>
+#include <strings/utilities.cuh>
+
+#include <algorithm>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+using translate_table = thrust::pair<char_utf8,char_utf8>;
+
+namespace
+{
+
+/**
+ * @brief This is the translate functor for replacing individual characters
+ * in each string.
+ */
+struct translate_fn
+{
+    column_device_view const d_strings;
+    translate_table const* table;
+    size_type table_size;
+    int32_t const* d_offsets{};
+    char* d_chars{};
+
+    __device__ size_type operator()(size_type idx)
+    {
+        if( d_strings.is_null(idx) )
+            return 0;
+        string_view d_str = d_strings.element<string_view>(idx);
+        size_type bytes = d_str.size_bytes();
+        char* out_ptr = nullptr;
+        if( d_offsets )
+            out_ptr = d_chars + d_offsets[idx];
+        for( auto chr : d_str )
+        {
+            auto entry = thrust::find_if( thrust::seq, table, table+table_size,
+                [chr] __device__ ( auto te ) { return te.first==chr; } );
+            if( entry < table + table_size )
+            {
+                bytes -= bytes_in_char_utf8(chr);
+                chr = entry->second;
+                if( chr ) // if null, remove character
+                    bytes += bytes_in_char_utf8(chr);
+            }
+            if( chr && out_ptr )
+                out_ptr += from_char_utf8(chr,out_ptr);
+        }
+        return bytes;
+    }
+};
+
+}
+
+//
+std::unique_ptr<column> translate( strings_column_view const& strings,
+                                   std::vector<std::pair<char_utf8,char_utf8>> const& chars_table,
+                                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                   cudaStream_t stream = 0 )
+{
+    size_type strings_count = strings.size();
+    if( strings_count == 0 )
+        return make_empty_strings_column(mr,stream);
+
+    size_type table_size = static_cast<size_type>(chars_table.size());
+    // convert input table
+    thrust::host_vector<translate_table> htable(table_size);
+    std::transform( chars_table.begin(), chars_table.end(), htable.begin(),
+        [] ( auto entry ) { return translate_table{entry.first,entry.second}; });
+    // copy translate table to device memory
+    rmm::device_vector<translate_table> table(htable);
+    translate_table* d_table = table.data().get();
```

Comment:
Is this necessary? What is motivating the `.data().get()`? It would be nice to modify `translate_fn` so we could avoid this if possible.

Reply:
Iterator approach means this is no longer needed.

Reply:
Awesome!